internal/addrs/parse_ref.go

```diff
@@ -396,13 +409,40 @@ func parseResourceRef(mode ResourceMode, startRange hcl.Range, traversal hcl.Tra
 	case hcl.TraverseAttr:
 		typeName = tt.Name
 	default:
-		// If it isn't a TraverseRoot then it must be a "data" reference.
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid reference",
-			Detail:   `The "data" object does not support this operation.`,
-			Subject:  traversal[0].SourceRange().Ptr(),
-		})
+		switch mode {
+		case ManagedResourceMode:
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid reference",
+				Detail:   `The "resource" object does not support this operation.`,
+				Subject:  traversal[0].SourceRange().Ptr(),
+			})
+		case DataResourceMode:
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid reference",
+				Detail:   `The "data" object does not support this operation.`,
+				Subject:  traversal[0].SourceRange().Ptr(),
+			})
+		case EphemeralResourceMode:
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid reference",
+				Detail:   `The "ephemeral" object does not support this operation.`,
+				Subject:  traversal[0].SourceRange().Ptr(),
+			})
+		default:
+			// Shouldn't get here because the above should be exhaustive for
+			// all of the resource modes. But we'll still return a
+			// minimally-passable error message so that the won't totally
+			// misbehave if we forget to update this in future.
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid reference",
+				Detail:   `The left operand does not support this operation.`,
+				Subject:  traversal[0].SourceRange().Ptr(),
+			})
+		}
```

Comment:
We can't right now, because the select block is exhaustive, but future changes could get caught here rather than passing through with something invalid.